gradio/blocks.py

```diff
@@ -2129,6 +2131,11 @@ def reverse(text):
 
         self.validate_queue_settings()
 
+        if self.dev_mode:
+            for block in self.blocks.values():
+                if block.key is None:
+                    block.key = f"__{block._id}__"
```

Comment:
Can we make the id a hash of the component's config? That way it will be unique across re-renders unless something about the component changed and we don't need to do the `reassign_keys` logic.

Reply:
So what you're suggesting is that we use the hash of the config to decide what components are considered identical across re-renders. While we could use that logic in `utils.reassign_keys` to consider what components are equivalent, it's gonna be problematic because it's very likely two components have the exact same config, e.g.

```python
with gr.Blocks():
  a = gr.Textbox()
  b = gr.Textbox()
```

I think it makes more sense to rely on the position and type of components to assign identity. The logic in `utils.reassign_keys` can always be improved but keep in mind this is only for dev mode refreshes - in actual render blocks, we will require users to explicitly assign keys to maintain identity.

Reply:
I see your point about duplicate keys but is that a concern for reload mode? Even if auto-generated ids collide, the component fetched in the frontend's `restored_keyed_values` would be correct since the config should determine the state of the component. 

The problem with using the position to determine equivalence is that if you insert a component "in between" or "before" already existing components, then the front-end state would get wiped.

This PR is already an improvement of what we currently do but wondering if we can create a better heuristic for equivalence from the get-go.

Reply:
> Even if auto-generated ids collide, the component fetched in the frontend's restored_keyed_values would be correct since the config should determine the state of the component.

I don't think this is the case. If initially I have:

```python
with gr.Blocks():
    boxA = gr.Textbox()
    boxB = gr.Textbox()
```
and I type something into boxA, and then I change boxA to:
```python
with gr.Blocks():
    boxA = gr.Textbox(label="A")
    boxB = gr.Textbox()
```
then the value that was in A would get moved to B, since that's the best match by config logic.

> This PR is already an improvement of what we currently do but wondering if we can create a better heuristic for equivalence from the get-go.

Sure, I think we can iterate on `reassign_keys` logic in other PRs, but since its only used in dev mode I think this can be a decent starting point.

Reply:
Ok sounds good!